File: compute/pyo3/python/mog/sub_apis/structure.py
"""Structural operations -- ``ws.structure.insert_rows()``, ``ws.structure.delete_columns()``."""
from __future__ import annotations

import json
import uuid
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union

from mog._serde import _col_to_a1, deserialize_mutation_result, parse_a1, parse_range
from mog.types import MutationResult

if TYPE_CHECKING:
    from mog._bridge import Bridge


def _unwrap_mutation(raw: Any) -> Any:
    """Extract the dict portion from a (bytes, dict) tuple if needed."""
    if isinstance(raw, tuple):
        for part in raw:
            if isinstance(part, dict):
                return part
    return raw


def _range_to_a1(sr: int, sc: int, er: int, ec: int) -> str:
    """Convert 0-based (sr, sc, er, ec) to A1 range string like 'A1:C3'."""
    return f"{_col_to_a1(sc)}{sr + 1}:{_col_to_a1(ec)}{er + 1}"


class StructureAPI:
    """Structural mutations (insert/delete rows/cols, merges) on a worksheet."""

    __slots__ = ("_bridge", "_sheet_id_json", "_protection_check")

    def __init__(self, bridge: Bridge, sheet_id_json: str, protection_check=None) -> None:
        self._bridge = bridge
        self._sheet_id_json = sheet_id_json
        self._protection_check = protection_check

    def _check_protection(self, op: str = "structural") -> None:
        """Raise if the sheet is protected and the operation is blocked."""
        if self._protection_check is not None and not self._protection_check(op):
            from mog.errors import MogError
            raise MogError(f"Cannot perform {op}: sheet is protected")

    # ------------------------------------------------------------------
    # Row / column insertion and deletion
    # ------------------------------------------------------------------

    def _sheet_id_hex(self) -> str:
        """Return the hex sheet ID (without JSON quotes)."""
        return json.loads(self._sheet_id_json)

    def insert_rows(self, at: int, count: int = 1) -> MutationResult:
        """Insert *count* rows at the given 0-based position.

        Returns a MutationResult with kind='insertRows'.
        """
        self._check_protection("insertRows")
        new_row_ids = [str(uuid.uuid4()) for _ in range(count)]
        change = {
            "InsertRows": {"at": at, "count": count, "new_row_ids": new_row_ids}
        }
        raw = self._bridge.structure_change(
            self._sheet_id_json, json.dumps(change)
        )
        result = deserialize_mutation_result(_unwrap_mutation(raw))
        # Enrich with structured receipt fields
        result.raw["kind"] = "insertRows"
        result.raw["sheetId"] = self._sheet_id_hex()
        result.raw["insertedAt"] = at
        result.raw["count"] = count
        return result

    def delete_rows(self, at: int, count: int = 1) -> MutationResult:
        """Delete *count* rows starting at the given 0-based position.

        Returns a MutationResult with kind='deleteRows'.
        """
        self._check_protection("deleteRows")
        change = {
            "DeleteRows": {"at": at, "count": count, "deleted_cell_ids": []}
        }
        raw = self._bridge.structure_change(
            self._sheet_id_json, json.dumps(change)
        )
        result = deserialize_mutation_result(_unwrap_mutation(raw))
        result.raw["kind"] = "deleteRows"
        result.raw["sheetId"] = self._sheet_id_hex()
        result.raw["deletedAt"] = at
        result.raw["count"] = count
        return result

    def insert_columns(self, at: int, count: int = 1) -> MutationResult:
        """Insert *count* columns at the given 0-based position.

        Returns a MutationResult with kind='insertColumns'.
        """
        self._check_protection("insertColumns")
        new_col_ids = [str(uuid.uuid4()) for _ in range(count)]
        change = {
            "InsertCols": {"at": at, "count": count, "new_col_ids": new_col_ids}
        }
        raw = self._bridge.structure_change(
            self._sheet_id_json, json.dumps(change)
        )
        result = deserialize_mutation_result(_unwrap_mutation(raw))
        result.raw["kind"] = "insertColumns"
        result.raw["sheetId"] = self._sheet_id_hex()
        result.raw["insertedAt"] = at
        result.raw["count"] = count
        return result

    def delete_columns(self, at: int, count: int = 1) -> MutationResult:
        """Delete *count* columns starting at the given 0-based position.

        Returns a MutationResult with kind='deleteColumns'.
        """
        self._check_protection("deleteColumns")
        change = {
            "DeleteCols": {"at": at, "count": count, "deleted_cell_ids": []}
        }
        raw = self._bridge.structure_change(
            self._sheet_id_json, json.dumps(change)
        )
        result = deserialize_mutation_result(_unwrap_mutation(raw))
        result.raw["kind"] = "deleteColumns"
        result.raw["sheetId"] = self._sheet_id_hex()
        result.raw["deletedAt"] = at
        result.raw["count"] = count
        return result

    # ------------------------------------------------------------------
    # Cell shift operations
    # ------------------------------------------------------------------

    def insert_cells_with_shift(
        self, sr: int, sc: int, er: int, ec: int, direction: str = "down"
    ) -> MutationResult:
        """Insert cells in a range and shift existing cells.

        Parameters
        ----------
        sr, sc, er, ec:
            0-based range coordinates.
        direction:
            ``"down"`` or ``"right"`` -- direction to shift existing cells.
        """
        shift_right = direction.lower() == "right"
        row_count = er - sr + 1
        col_count = ec - sc + 1
        raw = self._bridge.call_json(
            "compute_insert_cells_with_shift",
            self._sheet_id_json,
            sr, sc, row_count, col_count, shift_right,
        )
        result = deserialize_mutation_result(_unwrap_mutation(raw))
        result.raw["kind"] = "insertCells"
        result.raw["sheetId"] = self._sheet_id_hex()
        result.raw["range"] = {
            "startRow": sr, "startCol": sc, "endRow": er, "endCol": ec
        }
        result.raw["direction"] = direction
        return result

    def delete_cells_with_shift(
        self, sr: int, sc: int, er: int, ec: int, direction: str = "up"
    ) -> MutationResult:
        """Delete cells in a range and shift remaining cells.

        Parameters
        ----------
        sr, sc, er, ec:
            0-based range coordinates.
        direction:
            ``"up"`` or ``"left"`` -- direction to shift remaining cells.
        """
        shift_left = direction.lower() == "left"
        row_count = er - sr + 1
        col_count = ec - sc + 1
        raw = self._bridge.call_json(
            "compute_delete_cells_with_shift",
            self._sheet_id_json,
            sr, sc, row_count, col_count, shift_left,
        )
        result = deserialize_mutation_result(_unwrap_mutation(raw))
        result.raw["kind"] = "deleteCells"
        result.raw["sheetId"] = self._sheet_id_hex()
        result.raw["range"] = {
            "startRow": sr, "startCol": sc, "endRow": er, "endCol": ec
        }
        result.raw["direction"] = direction
        return result

    # ------------------------------------------------------------------
    # Remove duplicates
    # ------------------------------------------------------------------

    def remove_duplicates(
        self,
        range_str: str,
        columns: List[int],
        has_header: bool = False,
    ) -> Dict[str, Any]:
        """Remove duplicate rows from a range.

        Parameters
        ----------
        range_str:
            A1-style range string (e.g. ``"A1:B6"``).
        columns:
            List of 0-based column indices to check for duplicates.
        has_header:
            Whether the first row is a header row.

        Returns a dict with ``removedCount`` and ``remainingCount``.
        """
        sr, sc, er, ec = parse_range(range_str)
        raw = self._bridge.call_json(
            "compute_remove_duplicates",
            self._sheet_id_json,
            sr, sc, er, ec,
            json.dumps(columns),
            has_header,
        )
        # Extract data from the mutation result
        if isinstance(raw, dict):
            data = raw.get("data", {})
            if isinstance(data, dict):
                return {
                    "removedCount": data.get("duplicatesRemoved", 0),
                    "remainingCount": data.get("uniqueValuesRemaining", 0),
                }
        return {"removedCount": 0, "remainingCount": 0}

    # ------------------------------------------------------------------
    # Row / column counts
    # ------------------------------------------------------------------

    def get_row_count(self) -> int:
        """Return the number of rows containing data (uses data bounds)."""
        bounds = self._bridge.get_data_bounds(self._sheet_id_json)
        if bounds is None:
            return 0
        if isinstance(bounds, dict):
            max_row = bounds.get("maxRow", bounds.get("max_row", -1))
            if isinstance(max_row, (int, float)) and max_row >= 0:
                return int(max_row) + 1
        return 0

    def get_column_count(self) -> int:
        """Return the number of columns containing data (uses data bounds)."""
        bounds = self._bridge.get_data_bounds(self._sheet_id_json)
        if bounds is None:
            return 0
        if isinstance(bounds, dict):
            max_col = bounds.get("maxCol", bounds.get("max_col", -1))
            if isinstance(max_col, (int, float)) and max_col >= 0:
                return int(max_col) + 1
        return 0

    # ------------------------------------------------------------------
    # Merge operations
    # ------------------------------------------------------------------

    def merge(self, range_or_sr, sc=None, er=None, ec=None) -> MutationResult:
        """Merge cells in the given range.

        Accepts either an A1 range string (``"A1:C3"``) or four integer
        arguments ``(start_row, start_col, end_row, end_col)``.

        Returns a MutationResult with kind='merge' and range in A1 notation.
        """
        if sc is not None:
            sr, sc, er, ec = int(range_or_sr), int(sc), int(er), int(ec)
        else:
            sr, sc, er, ec = parse_range(range_or_sr)
        raw = self._bridge.merge_range(self._sheet_id_json, sr, sc, er, ec)
        result = deserialize_mutation_result(raw)
        result.raw["kind"] = "merge"
        result.raw["range"] = _range_to_a1(sr, sc, er, ec)
        return result

    def unmerge(self, range_or_sr, sc=None, er=None, ec=None) -> MutationResult:
        """Unmerge cells in the given range.

        Accepts either an A1 range string or four integer arguments.

        Returns a MutationResult with kind='unmerge' and range in A1 notation.
        """
        if sc is not None:
            sr, sc, er, ec = int(range_or_sr), int(sc), int(er), int(ec)
        else:
            sr, sc, er, ec = parse_range(range_or_sr)
        raw = self._bridge.unmerge_range(self._sheet_id_json, sr, sc, er, ec)
        result = deserialize_mutation_result(raw)
        result.raw["kind"] = "unmerge"
        result.raw["range"] = _range_to_a1(sr, sc, er, ec)
        return result

    def get_merged_regions(self) -> List[Any]:
        """Return a list of merged regions in this sheet."""
        result = self._bridge.call_json(
            "compute_get_all_merges_in_sheet", self._sheet_id_json
        )
        if isinstance(result, list):
            return result
        return []

    def get_merge_at_cell(
        self, address: Union[str, Tuple[int, int]]
    ) -> Optional[Dict[str, Any]]:
        """Get the merge region containing the given cell, or ``None``.

        Parameters
        ----------
        address:
            A1 address or ``(row, col)`` tuple.
        """
        if isinstance(address, tuple):
            row, col = address
        else:
            row, col = parse_a1(address)
        result = self._bridge.call_json(
            "compute_get_merge_at_cell_query", self._sheet_id_json, row, col
        )
        if isinstance(result, dict):
            return result
        return None

File: kernel/src/api/worksheet/structure.ts
/**
 * WorksheetStructureImpl — Implementation of the WorksheetStructure sub-API.
 *
 * Calls through to computeBridge directly for all structure and merge operations,
 * without the OperationResult/unwrap ceremony.
 */
import type {
  CellRange,
  DeleteCellsReceipt,
  DeleteColumnsReceipt,
  DeleteRowsReceipt,
  InsertCellsReceipt,
  InsertColumnsReceipt,
  InsertRowsReceipt,
  MergeReceipt,
  MergedRegion,
  RemoveDuplicatesResult,
  SheetId,
  TextToColumnsOptions,
  TextToColumnsResult,
  UnmergeReceipt,
  WorksheetStructure,
} from '@mog-sdk/contracts/api';

import type { ProtectionOperation } from '@mog-sdk/contracts/api';

import type { MutationResult } from '../../bridges/compute/compute-types.gen';
import type { DocumentContext } from '../../context';
import { KernelError } from '../../errors';
import { resolveCell, resolveRange } from '../internal/address-resolver';
import { normalizeRange, toA1 } from '../internal/utils';
import { WorksheetProtectionImpl } from './protection';

export class WorksheetStructureImpl implements WorksheetStructure {
  private _protection?: WorksheetProtectionImpl;

  constructor(
    private readonly ctx: DocumentContext,
    private readonly sheetId: SheetId,
  ) {}

  private get protectionApi(): WorksheetProtectionImpl {
    return (this._protection ??= new WorksheetProtectionImpl(this.ctx, this.sheetId));
  }

  /**
   * Guard: throws WriteGateError if the document is not writable (the write gate).
   */
  private _ensureWritable(operation: string): void {
    this.ctx.writeGate.assertWritable(operation);
  }

  /**
   * Throws if the sheet is protected and the given structure operation is not allowed.
   */
  private async ensureStructureOpAllowed(operation: ProtectionOperation): Promise<void> {
    const allowed = await this.protectionApi.canDoStructureOp(operation);
    if (!allowed) {
      throw new KernelError(
        'OPERATION_FAILED',
        `Cannot perform ${operation}: sheet is protected and operation is not allowed`,
      );
    }
  }

  // ===========================================================================
  // Row / Column insertion and deletion
  // ===========================================================================

  async insertRows(index: number, count: number): Promise<InsertRowsReceipt> {
    this._ensureWritable('structure.insertRows');
    if (index < 0) throw new KernelError('API_INVALID_ADDRESS', `Invalid row index: ${index}`);
    if (count <= 0)
      return { kind: 'insertRows', sheetId: this.sheetId, insertedAt: index, count: 0 };
    await this.ensureStructureOpAllowed('insertRows');
    const result = (await this.ctx.computeBridge.structureChange(this.sheetId, {
      InsertRows: { at: index, count, new_row_ids: [] },
    })) as MutationResult | void;
    const sc = result?.structureChanges?.[0];
    return {
      kind: 'insertRows',
      sheetId: sc?.sheetId ?? this.sheetId,
      insertedAt: sc?.at ?? index,
      count: sc?.count ?? count,
    };
  }

  async deleteRows(index: number, count: number): Promise<DeleteRowsReceipt> {
    this._ensureWritable('structure.deleteRows');
    if (index < 0) throw new KernelError('API_INVALID_ADDRESS', `Invalid row index: ${index}`);
    if (count <= 0)
      return { kind: 'deleteRows', sheetId: this.sheetId, deletedAt: index, count: 0 };
    await this.ensureStructureOpAllowed('deleteRows');
    const result = (await this.ctx.computeBridge.structureChange(this.sheetId, {
      DeleteRows: { at: index, count, deleted_cell_ids: [] },
    })) as MutationResult | void;
    const sc = result?.structureChanges?.[0];
    return {
      kind: 'deleteRows',
      sheetId: sc?.sheetId ?? this.sheetId,
      deletedAt: sc?.at ?? index,
      count: sc?.count ?? count,
    };
  }

  async insertColumns(index: number, count: number): Promise<InsertColumnsReceipt> {
    this._ensureWritable('structure.insertColumns');
    if (index < 0) throw new KernelError('API_INVALID_ADDRESS', `Invalid column index: ${index}`);
    if (count <= 0)
      return { kind: 'insertColumns', sheetId: this.sheetId, insertedAt: index, count: 0 };
    await this.ensureStructureOpAllowed('insertColumns');
    const result = (await this.ctx.computeBridge.structureChange(this.sheetId, {
      InsertCols: { at: index, count, new_col_ids: [] },
    })) as MutationResult | void;
    const sc = result?.structureChanges?.[0];
    return {
      kind: 'insertColumns',
      sheetId: sc?.sheetId ?? this.sheetId,
      insertedAt: sc?.at ?? index,
      count: sc?.count ?? count,
    };
  }

  async deleteColumns(index: number, count: number): Promise<DeleteColumnsReceipt> {
    this._ensureWritable('structure.deleteColumns');
    if (index < 0) throw new KernelError('API_INVALID_ADDRESS', `Invalid column index: ${index}`);
    if (count <= 0)
      return { kind: 'deleteColumns', sheetId: this.sheetId, deletedAt: index, count: 0 };
    await this.ensureStructureOpAllowed('deleteColumns');
    const result = (await this.ctx.computeBridge.structureChange(this.sheetId, {
      DeleteCols: { at: index, count, deleted_cell_ids: [] },
    })) as MutationResult | void;
    const sc = result?.structureChanges?.[0];
    return {
      kind: 'deleteColumns',
      sheetId: sc?.sheetId ?? this.sheetId,
      deletedAt: sc?.at ?? index,
      count: sc?.count ?? count,
    };
  }

  // ===========================================================================
  // Cell shifting
  // ===========================================================================

  async insertCellsWithShift(
    startRow: number,
    startCol: number,
    endRow: number,
    endCol: number,
    direction: 'right' | 'down',
  ): Promise<InsertCellsReceipt> {
    this._ensureWritable('structure.insertCellsWithShift');
    await this.ensureStructureOpAllowed(direction === 'right' ? 'insertColumns' : 'insertRows');
    const rowCount = endRow - startRow + 1;
    const colCount = endCol - startCol + 1;
    const shiftRight = direction === 'right';
    await this.ctx.computeBridge.insertCellsWithShift(
      this.sheetId,
      startRow,
      startCol,
      rowCount,
      colCount,
      shiftRight,
    );
    return {
      kind: 'insertCells',
      sheetId: this.sheetId,
      range: { startRow, startCol, endRow, endCol },
      direction,
    };
  }

  async deleteCellsWithShift(
    startRow: number,
    startCol: number,
    endRow: number,
    endCol: number,
    direction: 'left' | 'up',
  ): Promise<DeleteCellsReceipt> {
    this._ensureWritable('structure.deleteCellsWithShift');
    await this.ensureStructureOpAllowed(direction === 'left' ? 'deleteColumns' : 'deleteRows');
    const rowCount = endRow - startRow + 1;
    const colCount = endCol - startCol + 1;
    const shiftLeft = direction === 'left';
    await this.ctx.computeBridge.deleteCellsWithShift(
      this.sheetId,
      startRow,
      startCol,
      rowCount,
      colCount,
      shiftLeft,
    );
    return {
      kind: 'deleteCells',
      sheetId: this.sheetId,
      range: { startRow, startCol, endRow, endCol },
      direction,
    };
  }

  // ===========================================================================
  // Dimensions
  // ===========================================================================

  async getRowCount(): Promise<number> {
    try {
      const bounds = await this.ctx.computeBridge.getDataBounds(this.sheetId);
      if (!bounds) return 0;
      return bounds.maxRow + 1;
    } catch {
      return 0;
    }
  }

  async getColumnCount(): Promise<number> {
    try {
      const bounds = await this.ctx.computeBridge.getDataBounds(this.sheetId);
      if (!bounds) return 0;
      return bounds.maxCol + 1;
    } catch {
      return 0;
    }
  }

  // ===========================================================================
  // Data operations
  // ===========================================================================

  async textToColumns(
    range: string | CellRange,
    options: TextToColumnsOptions,
  ): Promise<TextToColumnsResult> {
    this._ensureWritable('structure.textToColumns');
    const parsed = resolveRange(range);
    const dest = resolveTextToColumnsDestination(options.destination, parsed);
    const bridgeOptions = toBridgeTextToColumnsOptions(options);
    const raw = await this.ctx.computeBridge.textToColumns(
      this.sheetId,
      parsed.startRow,
      parsed.endRow,
      parsed.startCol,
      dest.row,
      dest.col,
      bridgeOptions,
    );
    // The ComputeBridge.textToColumns override handles the YRS → mirror sync
    // (read-back + setCellsByPosition) so no forceRefreshAllViewports needed here.
    const payload = (raw as MutationResult).data as
      | { rowsProcessed?: number; columnsCreated?: number }
      | undefined;
    return {
      rowsProcessed: payload?.rowsProcessed ?? parsed.endRow - parsed.startRow + 1,
      columnsCreated: payload?.columnsCreated ?? 1,
    };
  }

  async removeDuplicates(
    range: string | CellRange,
    columns: number[],
    hasHeaders?: boolean,
  ): Promise<RemoveDuplicatesResult> {
    this._ensureWritable('structure.removeDuplicates');
    const parsed = resolveRange(range);
    const raw = await this.ctx.computeBridge.removeDuplicates(
      this.sheetId,
      parsed.startRow,
      parsed.startCol,
      parsed.endRow,
      parsed.endCol,
      columns,
      hasHeaders ?? false,
    );
    // Rust returns full-viewport patches for this operation and the viewport
    // coordinator applies those patches directly, so no follow-up refresh is
    // needed here.
    // The bridge returns MutationResult whose `.data` field carries the
    // removeDuplicates stats as a JSON object set via `with_data` on the Rust side.
    const payload = (raw as MutationResult).data as
      | { duplicatesRemoved?: number; uniqueValuesRemaining?: number }
      | undefined;
    return {
      removedCount: payload?.duplicatesRemoved ?? 0,
      remainingCount: payload?.uniqueValuesRemaining ?? 0,
    };
  }

  // ===========================================================================
  // Merges
  // ===========================================================================

  async merge(
    a: string | number | CellRange,
    b?: number,
    c?: number,
    d?: number,
  ): Promise<MergeReceipt> {
    this._ensureWritable('structure.merge');
    const bounds = resolveRange(a, b, c, d);
    const normalized = normalizeRange(bounds);

    if (normalized.startRow === normalized.endRow && normalized.startCol === normalized.endCol) {
      throw new KernelError('COMPUTE_ERROR', 'Cannot merge a single cell');
    }

    await this.ctx.computeBridge.mergeRange(
      this.sheetId,
      normalized.startRow,
      normalized.startCol,
      normalized.endRow,
      normalized.endCol,
    );
    // Merge changes cell geometry — force-refresh all viewports so the
    // ViewportMergeIndex is rebuilt immediately (not deferred to next scroll).
    // Without this, clicking a merged slave cell right after the merge
    // finds a stale VMI (null entry), so the active cell snaps to the
    // clicked col rather than the merge's top-left origin.
    this.ctx.computeBridge.invalidateAllViewportPrefetch();
    await this.ctx.computeBridge.forceRefreshAllViewports();
    return {
      kind: 'merge',
      range: `${toA1(normalized.startRow, normalized.startCol)}:${toA1(normalized.endRow, normalized.endCol)}`,
    };
  }

  async unmerge(
    a: string | number | CellRange,
    b?: number,
    c?: number,
    d?: number,
  ): Promise<UnmergeReceipt> {
    this._ensureWritable('structure.unmerge');
    const bounds = resolveRange(a, b, c, d);
    const normalized = normalizeRange(bounds);

    const overlappingMerges = await this.ctx.computeBridge.getMergesInViewportSpatial(
      this.sheetId,
      normalized.startRow,
      normalized.startCol,
      normalized.endRow,
      normalized.endCol,
    );
    if (overlappingMerges.length === 0) {
      return {
        kind: 'unmerge',
        range: `${toA1(normalized.startRow, normalized.startCol)}:${toA1(normalized.endRow, normalized.endCol)}`,
      };
    }

    await this.ctx.computeBridge.unmergeRange(
      this.sheetId,
      normalized.startRow,
      normalized.startCol,
      normalized.endRow,
      normalized.endCol,
    );
    // Unmerge changes cell geometry — force-refresh all viewports so the
    // ViewportMergeIndex is rebuilt immediately (not deferred to next scroll).
    this.ctx.computeBridge.invalidateAllViewportPrefetch();
    await this.ctx.computeBridge.forceRefreshAllViewports();
    return {
      kind: 'unmerge',
      range: `${toA1(normalized.startRow, normalized.startCol)}:${toA1(normalized.endRow, normalized.endCol)}`,
    };
  }

  async getMergedRegions(): Promise<MergedRegion[]> {
    const regions = await this.ctx.computeBridge.getAllMergesInSheet(this.sheetId);
    return regions.map((r) => ({
      range: `${toA1(r.startRow, r.startCol)}:${toA1(r.endRow, r.endCol)}`,
      startRow: r.startRow,
      startCol: r.startCol,
      endRow: r.endRow,
      endCol: r.endCol,
      rowSpan: r.endRow - r.startRow + 1,
      colSpan: r.endCol - r.startCol + 1,
    }));
  }

  async getMergeAtCell(a: string | number, b?: number): Promise<CellRange | null> {
    const { row, col } = resolveCell(a, b);
    const info = await this.ctx.computeBridge.getMergeAtCellQuery(this.sheetId, row, col);
    if (!info) return null;
    return {
      startRow: info.merge.startRow,
      startCol: info.merge.startCol,
      endRow: info.merge.endRow,
      endCol: info.merge.endCol,
    };
  }
}

// =============================================================================
// Internal helpers
// =============================================================================

function resolveTextToColumnsDestination(
  destination: TextToColumnsOptions['destination'],
  source: CellRange,
): { row: number; col: number } {
  if (!destination) {
    return { row: source.startRow, col: source.startCol };
  }
  if (typeof destination === 'string') {
    return resolveCell(destination);
  }
  if (
    Number.isInteger(destination.row) &&
    destination.row >= 0 &&
    Number.isInteger(destination.col) &&
    destination.col >= 0
  ) {
    return destination;
  }
  throw new KernelError(
    'API_INVALID_ADDRESS',
    `Invalid text-to-columns destination: ${JSON.stringify(destination)}`,
  );
}

function toBridgeTextToColumnsOptions(options: TextToColumnsOptions): Record<string, unknown> {
  const splitType = options.type === 'fixedWidth' ? 'fixedWidth' : 'delimited';
  let delimiters: Record<string, boolean | string | undefined>;
  if (options.delimiters) {
    delimiters = {
      tab: options.delimiters.tab ?? false,
      comma: options.delimiters.comma ?? false,
      semicolon: options.delimiters.semicolon ?? false,
      space: options.delimiters.space ?? false,
    };
    if (options.delimiters.other) {
      delimiters.other = options.delimiters.other;
    }
  } else {
    const delimiter = options.delimiter ?? 'comma';
    delimiters = {
      tab: delimiter === 'tab',
      comma: delimiter === 'comma',
      semicolon: delimiter === 'semicolon',
      space: delimiter === 'space',
    };
    if (delimiter === 'custom' && options.customDelimiter) {
      delimiters.other = options.customDelimiter;
    }
  }

  let textQualifier: string;
  if (options.textQualifier === "'") {
    textQualifier = 'singleQuote';
  } else if (options.textQualifier === 'none') {
    textQualifier = 'none';
  } else {
    // default: double-quote (matches Rust default)
    textQualifier = 'doubleQuote';
  }

  return {
    splitType,
    delimiters,
    treatConsecutiveAsOne: options.treatConsecutiveAsOne ?? false,
    textQualifier,
    fixedWidthBreaks: options.fixedWidthBreaks ?? [],
  };
}
